### ISSUE TITLE
Address spaces not listed for pointer argument of wait_group_events()

### ISSUE THREAD

**Issue by StuartDBrady** [KhronosGroup/OpenCL-Docs#472]
The description of **wait_group_events()** in the unified OpenCL specification does not list the valid address spaces for the `event_t *` pointer argument of this function.

For OpenCL _1.x_, and for OpenCL 3 _without_ the generic address space feature, it seems clear that this function takes a pointer to the private address space.

For OpenCL _2.x_ and for OpenCL 3 _with_ the generic address space feature, it is clear that pointers to the private address space must still be accepted, but it is not clear whether the pointers to the generic address space should also be accepted.

What are the required address spaces that must be supported for this function's pointer argument?

**Comment by alycm** [comment]
> For OpenCL 1.x, and for OpenCL 3 without the generic address space feature, it seems clear that this function takes a pointer to the private address space.

Agreed.

> For OpenCL 2.x and for OpenCL 3 with the generic address space feature, it is clear that pointers to the private address space must still be accepted, but it is not clear whether the pointers to the generic address space should also be accepted.

For OpenCL C 2.0 it seems easy to construct a program that uses a generic pointer to an event_t and passes that to `wait_group_events` so I would say that "yes, pointers to the generic address space should also be accepted" for these cases.

So I think that the fix is to add `__private` to the pointer explicitly in the specification and also a generic version with wording such as:

> For OpenCL C 2.0, or OpenCL C 3.0 or newer with the `__opencl_c_generic_address_space` feature:

Of course, this need considered carefully to ensure that it doesn't represent a functional change.  I think the fact that `event_t` can only exist in the private address space limits helps here.  I.e. the [OpenCL C restrictions section](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_C.html#restrictions) says:

> The event type (`event_t`) cannot be used as the type of a kernel function argument. The event type cannot be used to declare a program scope variable. The event type cannot be used to declare a structure or union field. The event type cannot be used with the `__local`, `__constant` and `__global` address space qualifiers.

**Comment by bashbaug** [comment]
This reminds me a little of the discussions we had regarding pipes and device-side enqueue (internal issues 235 and 236).

Since `wait_group_events` was originally an OpenCL 1.x API and not an OpenCL 2.x API I agree we can _probably_ just add the private address space overload and a generic address space overload when the generic address space is supported (OpenCL C 2.0, or OpenCL C 3.0 or newer with the `__opencl_c_generic_address_space` feature) and be done with it.

I did a quick search through the CTS source and I only ever see `wait_group_events` being called with a pointer to the private address space, for whatever that's worth.

**Comment by StuartDBrady** [comment]
True, but the prototype for *wait_group_events()* in Clang's {{opencl-c.h}} does not specify an address space qualifier, and so when compiling using upstream Clang, the CTS's calls to *wait_group_events()* will use the generic address space when compiling for OpenCL 2.0.

I am unsure if there has been a recent change to the CTS that would the *wait_group_events()* test to be compiled using {{-cl-std=CL2.0}}.

Arm's implementation does not provide a version of *wait_group_events()* using the generic address space. It is not clear whether it should.

In SPIR-V, it is not clear to me what storage classes are permissible for the pointer argument of *OpGroupWaitEvents*, but it seems the _Khronos SPIR-V/LLVM Translator_ translates *OpGroupWaitEvents* instructions using the *Generic* storage class into *wait_group_events()* calls using the generic address space.

**Comment by AnastasiaStulova** [comment]
Summary from the teleconference on Feb 11, 2021
- conceptually only private address space makes sense as events can only be allocated in private memory but this was missing in OpenCL 2.0 because the private qualifier was committed. 
- It seems the safest approach is to add generic address space overload if the generic feature is supported.

**Comment by AnastasiaStulova** [comment]
Summary from the teleconference on Mar 11, 2021
- This needs PR for the spec and might be a bug to Clang.
